scipy/linalg/special_matrices.py

```diff
@@ -1194,3 +1194,82 @@ def fiedler_companion(a):
     c[[0, 1], 0] = [-a[1], 1]
 
     return c
+
+def convmtx(a, n, mode='full'):
+    """
+    Construct a convolution matrix.
+
+    Constructs the dense matrix representing convolution[1].
+
+    `A = convmtx(a, n[, mode])`
+    creates a matrix `A` such that
+    `np.dot(A, v)` is equivalent to `convolve(a, v[, mode])` but slower.
+    In the default 'full' mode, an element
+    `A[i,j] == a[i-j]`, if `a[i-j]` exists, otherwise zero.
+
+    Parameters
+    ----------
+    a : array
+        The array to convolve.
+    n : int
+        The number of columns in the resulting matrix.
+        This is analogous to the length of `v` in `numpy.convolve(a,v)`
+    mode : string
+        This is analogous to `mode` in numpy.convolve(v, a, mode).
+        It determines the number of rows in A as
+
+            - 'full'[default]: `len(a) + n - 1`
+            - 'same': `max(n, len(a) )`
+            - 'valid': `max(n, len(a) ) - min(n, len(a) ) + 1`
+
+    Returns
+    -------
+    A : The convolution matrix
+
+    Notes
+    -----
+    .. versionadded:: 1.4.0
+
+    See Also
+    --------
+    toeplitz : Toeplitz matrix
+
+    References
+    ----------
+    .. [1] "Convolution", https://en.wikipedia.org/wiki/Convolution
+
+    Examples
+    --------
+    >>> from scipy.linalg import convmtx
+    >>> convmtx( (-1,2,-1), 5, mode='same')
+    array([[ 2, -1,  0,  0,  0],
+           [-1,  2, -1,  0,  0],
+           [ 0, -1,  2, -1,  0],
+           [ 0,  0, -1,  2, -1],
+           [ 0,  0,  0, -1,  2]])
+
+    """
+    a = np.asarray(a)
+    if a.ndim != 1:
+        raise ValueError('convmtx expects a 1d array as input')
+
```

Comment:
Here we need a validation of the mode argument via say
```
    if mode not in ['full', 'same', 'valid']:
        raise ValueError("'mode' argument must be one of the 'full', 'same',"
                         " 'valid'")
```
Then we can safely use the `else` block below.

Reply:
mode restricted as of 2a1bd974e5605d103028f8b64919d30ba900aa7d